scipy/sparse/tests/test_base.py

```diff
@@ -731,6 +731,9 @@ def test_diagonal(self):
 
         # Test all-zero matrix.
         assert_equal(self.spmatrix((40, 16130)).diagonal(), np.zeros(40))
+        # Test empty matrix
+        # https://github.com/scipy/scipy/issues/11949
+        assert_equal(self.spmatrix((0, 0)).diagonal(), np.empty(0))
```

Comment:
Can you include all the test cases from gh-11949 here?

Reply:
Below test classes inherited from ```_TestCommon```, so test running for all existing classes(except `NonCannonicalBSR`)

If double test would be OK, the I'll add test from #11949

Reply:
Sorry, I wasn't specific enough. What I meant to ask for was including Warren's additional cases, that cover the following shapes and offsets:

```
In [120]: np.diagonal(np.zeros((3, 5)), offset=99)
Out[120]: array([], dtype=float64)

In [121]: np.diagonal(np.zeros((3, 0)), offset=0)
Out[121]: array([], dtype=float64)

In [122]: np.diagonal(np.zeros((0, 0)), offset=0)
Out[122]: array([], dtype=float64)
```

To keep things simple, you can achieve this by asserting that the result of the `diagonal()` method is the same for sparse matrix `s` and the equivalent numpy array `s.A`.

Reply:
Got it.
In the last commit added tests to cover these cases